sled-agent/src/artifact_store.rs

```diff
@@ -514,16 +535,18 @@ impl ArtifactWriter {
 
         self.files.retain(Option::is_some);
         if self.files.is_empty() {
-            Err(last_error.unwrap_or(Error::NoUpdateDataset))
+            Err(self.last_error.take().unwrap_or(Error::NoUpdateDataset))
         } else {
             Ok(())
         }
     }
 
-    /// Rename all files to their final paths. If an error occurs, it is logged.
-    /// If none of the files are renamed successfully, the most recently-seen
-    /// error is returned.
-    async fn finalize(self) -> Result<(), Error> {
+    /// Rename all files to their final paths.
+    ///
+    /// Errors in this method are considered non-fatal errors, but this method
+    /// will return the most recently-seen error by any method in the write
+    /// process.
```

Comment:
FWIW one possible solution here would be to propagate a result back through the API, that indicates:

- "We wrote to two M.2s successfully"
- "We wrote to one M.2 successfully, and the other failed (here's the error)"
- "Neither write completed successfully, here are the errors (or the most recent error)"

Then this decision is punted up to Nexus, and Nexus could decide "at least one write count as success, but I'll log the error and keep moving".

Reply:
Nexus could call the `list` API to see if there was a partial success, I suppose; or maybe it's worth instead returning something like `{"datasets": 2, "successful_writes": 1}` as a non-4xx error?

Reply:
I think the actual answer to this question depends on the exact design of how Nexus is going to replicate artifacts across sleds. If Nexus is able to try another sled, maybe this current design is fine. But if it's a saga where the sleds are picked out in advance and there's no retry flow (this seems like a poor design) then it would be better to return OK here; at least there's a copy on this one sled.

Reply:
In a hypothetical world where we have all sleds operating with one M.2 - shouldn't this be able to succeed? We are operating at reduced redundancy, but we *do* have a copy that got successfully written.

(Agreed that we could make all `PUT` calls query the list API afterwards? But if that's our inclination, this also seems like it should be part of the error result)

basically, I think it's critical for Nexus to be able to distinguish between the cases of:
- We successfully wrote to at least one one M.2, and
- Every other possible outcome

Reply:
I'll implement the `{"datasets": 2, "successful_writes": 1}` 200 OK response so that Nexus can make a decision with that information. I don't think it matters to return the error since sled agent is logging all I/O errors it runs into.